openvdb/openvdb/points/PointStatistics.h

```diff
@@ -0,0 +1,611 @@
+    tbb::parallel_for(manager.leafRange(),
+        [idx, &filter, &values] (const auto& range) {
+            for (auto leaf = range.begin(); leaf; ++leaf) {
+                AttributeHandle<ValueT, CodecT> handle(leaf->constAttributeArray(idx));
+                size_t size = handle.size();
+                if (size == 0) continue;
+                if (std::is_same<FilterT, NullFilter>::value) {
+                    if (handle.isUniform()) size = 1;
+                    ResultT total = ResultT(handle.get(0));
+                    for (size_t i = 1; i < size; ++i) {
+                        assert(i < size_t(std::numeric_limits<Index>::max()));
+                        total += ResultT(handle.get(Index(i)));
+                    }
+                    // ignore int/float warnings
+                    OPENVDB_NO_TYPE_CONVERSION_WARNING_BEGIN
+                    values[leaf.pos()].reset(new ResultT(total / DivisorT(size)));
+                    OPENVDB_NO_TYPE_CONVERSION_WARNING_END
+                }
+                else {
+                    auto iter = leaf->beginIndexOn(filter);
+                    if (!iter) continue;
+                    ResultT total = ResultT(handle.get(*iter));
+                    ++iter;
+                    size = 1;
+                    for (; iter; ++iter, ++size) total += ResultT(handle.get(*iter));
+                    // ignore int/float warnings
+                    OPENVDB_NO_TYPE_CONVERSION_WARNING_BEGIN
+                    values[leaf.pos()].reset(new ResultT(total / DivisorT(size)));
+                    OPENVDB_NO_TYPE_CONVERSION_WARNING_END
+                }
+            }
+        });
+
+    auto iter = values.cbegin();
+    while (!(*iter) && iter != values.cend()) ++iter;
+    if (iter == values.cend()) return false;
+    assert(*iter);
+
+    average = **iter; ++iter;
+    uint_fast64_t count = 1;
+
+    if (std::is_integral<ElementT>::value) {
+        using RangeT = tbb::blocked_range<const std::unique_ptr<ResultT>*>;
+        using PairT = std::pair<ResultT, uint_fast64_t>;
+        PairT tc = { average, 1 };
+        // reasonable grain size for accumulation of single to vec/matrix types
+        tc = tbb::parallel_reduce(RangeT(&(*iter), &values.back(), 32), tc,
+           [](const RangeT& range, PairT p) -> PairT {
+                for (const auto& r : range) {
+                    if (!r) continue;
+                    p.first += *r;
+                    ++p.second;
+                }
+                return p;
+           }, [](PairT a, const PairT& b) -> PairT {
+                a.first += b.first;
+                a.second += b.second;
+                return a;
+           });
+
+        average = tc.first;
+        count = tc.second;
+    }
+    else {
+        for (; iter != values.cend(); ++iter) {
+            if (!*iter) continue;
+            average += (**iter);
+            ++count;
+        }
+    }
+
+    average = ResultT(average / static_cast<DivisorT>(count));
```

Comment:
The way this works if I have read it correctly, it calcs the average per leaf, stores that the in values vector and then divides by the number of leaves (i.e. non null elements in values) right? If so I dont think that works for averaging. I.e. if a leaf had one point of 6 with attribute that was 1, the rest 0 (i.e. values[idx] = 1/6) and then another leaf had one point that was 1 and the rest 0, from 2 points (i.e. values[otheridx] = 1/2), then this assumes (1/6 + 1/2)/2 == 2/8 but that isnt true. (1/6 + 1/2)/2 = 1/3. I assume this isnt picked up as the tests all use the same number of active points in each leaf?

Reply:
This was a silly mistake thanks for picking this up @richhones. Fixed